contracts/verifiers/UniversalVerifierMultiQuery.sol

```diff
@@ -0,0 +1,801 @@
+        private
+        pure
+        returns (UniversalVerifierMultiQueryStorage storage $)
+    {
+        assembly {
+            $.slot := UniversalVerifierMultiQueryStorageLocation
+        }
+    }
+
+    /**
+     * @dev Sets a request
+     * @param requestId The ID of the request
+     * @param request The request data
+     */
+    function setRequest(
+        uint256 requestId,
+        Request calldata request
+    )
+        public
+        checkRequestExistence(requestId, false)
+        checkRequestGroupExistence(request.validator.getGroupID(request.params), false)
+        onlyWhitelistedValidator(request.validator)
+    {
+        UniversalVerifierMultiQueryStorage storage s = _getUniversalVerifierMultiQueryStorage();
+        s._requests[requestId] = request;
+        s._requestIds.push(requestId);
+
+        emit RequestSet(
+            requestId,
+            _msgSender(),
+            request.metadata,
+            address(request.validator),
+            request.params
+        );
+    }
+
+    /**
+     * @dev Sets a group of requests
+     * @param groupId The ID of the group
+     * @param requestIds The IDs of the requests
+     * @param requests The requests data
+     */
+    function setGroupOfRequests(
+        uint256 groupId,
+        uint256[] memory requestIds,
+        Request[] calldata requests
+    ) public {
+        UniversalVerifierMultiQueryStorage storage s = _getUniversalVerifierMultiQueryStorage();
+
+        for (uint256 i = 0; i < requestIds.length; i++) {
+            setRequest(requestIds[i], requests[i]);
+        }
+        // check that all the requests are set before adding them to the group
+        // because there is a check in setRequest that the group doesn't exist yet
+        for (uint256 i = 0; i < requestIds.length; i++) {
+            s._groupedRequests[groupId].push(requestIds[i]);
+        }
+    }
+
+    /**
+     * @dev Gets a specific request by ID
+     * @param requestId The ID of the request
+     * @return request The request data
+     */
+    function getRequest(
+        uint256 requestId
+    ) public view checkRequestExistence(requestId, true) returns (Request memory request) {
+        return _getUniversalVerifierMultiQueryStorage()._requests[requestId];
+    }
+
+    /**
+     * @dev Checks auth request in query
+     * @param query The query data
+     */
+    function _checkAuthRequestInQuery(Query calldata query) internal pure {
+        uint256 numAuthRequests = 0;
+
+        for (uint256 i = 0; i < query.requestIds.length; i++) {
+            if (_getRequestType(query.requestIds[i]) == AUTH_REQUEST_TYPE) {
+                numAuthRequests++;
+            }
+        }
+
+        require(numAuthRequests == 1, "Exactly 1 auth request is required");
+    }
+
+    /**
+     * @dev Sets a query
+     * @param queryId The ID of the query
+     * @param query The query data
+     */
+    function setQuery(
+        uint256 queryId,
+        Query calldata query
```

Comment:
As an idea:
What if the Query would have two arrays `requestIds[]` and `groupIds[]`? You will check that each element in the `requestIds` array has no group assigned.
That should simplify the logic I guess and you can remove the `_checkRequestsInGroupsIncluded`